File: day16/.py
def pack_gifts(gifts: list[int], max_weight: int) -> int | None:
  # ps: Ik the original arg was maxWeight but I j made it more pythronic
  if not len(gifts): return 0
  sleigh_count = 1
  current_weight = 0

  for gift_weight in gifts:
    if (gift_weight > max_weight): return None
    if gift_weight + current_weight > max_weight:
      current_weight = gift_weight
      sleigh_count += 1
    else:
      current_weight += gift_weight
  return sleigh_count

File: day16/.js
/**
 * @param {number[]} gifts - The gifts to pack
 * @param {number} maxWeight - The maximum weight of the sleigh
 * @returns {number | null} The number of sleighs needed
 * Return null if no sleigh can carry all the gifts
 */
function packGifts(gifts, maxWeight) {
  if (!gifts.length) return 0;
  let sleighCount = 1;
  let currentWeight = 0;

  for (let giftWeight of gifts){
    if (giftWeight > maxWeight) return null 
    if (giftWeight + currentWeight > maxWeight){
      currentWeight = giftWeight
      sleighCount += 1
    } else {
      currentWeight += giftWeight
    }
  }
  return sleighCount
}
